417 - Add edit and reorder metadata within galleryView modal

dstopp opened the pull request:
- Fixes #417
- Were any Django, import pipeline, table_schema, or dictionary files modified? Y/N
  - Database used: opus3_test_200517
  - All Django tests pass: Y/N/NA
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y
  - Tested on Chrome, Firefox, Safari, and iOS: Y
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): Y/N/NA

Description of changes:
-add an 'edit' button on the galleryView modal to allow user to:
   - add a metadata field from the menu list;
   - remove a metadata field;
   - reorder the metadata fields
-Fixed race condition between search widget and selectMetadata modal that cause unselected fields to be highlighted
-Fixed missing scrollbar <bug> in modal

Known problems:
-There is some code that is unused in this pull that will eventually be part of the table header draggable code.  I did not remove it for this pull.
-There are still 3 django tests failing that appear to not be related to these changes but are still under investigation.

dstopp committed:
sandbox

dstopp committed:
keep the transitioning column width the same as the header is moved

dstopp committed:
add an opacity transition while the columns are being manipulated;
maintain both header and cell width during transition so it doesn't look odd until the table is redrawn

dstopp committed:
attempt to make the column move smoother by maintaining width

dstopp committed:
this looks better; header still drag w/out column but extra column doesn't seem to appear

dstopp committed:
more test code

dstopp committed:
really darken when header moves

dstopp committed:
more minor changes

dstopp committed:
more test code added

dstopp committed:
-allow detail window <slide show view> metadata details to be reordered
-minor changes for column reording <not done>

dstopp committed:
minor cosmetic fix

dstopp committed:
add a class to the marked menu items so that they can be later filtered for the marked items

dstopp committed:
add edit icons and mini menu for adding selected metadata items

dstopp committed:
fix bizarre copy and paste error

dstopp committed:
more updates

juzen2003 committed:
Moving data-toggle, aria-haspopup and aria-expanded to .op-metadata-detail-add
a tag button. Separate these attributes from .dropdown-menu
(#op-add-metadata-fields) to avoid the weird behavior that its sibling
.container-fluid keeps toggling show class when any collapsible category
inside #op-add-metadata-fields.

juzen2003 committed:
Remove unwanted console.log

dstopp committed:
Merge pull request #6 from juzen2003/417_reorder_table_column

417 reorder table column

dstopp committed:
-highlight and update after add of a new metadata field
-unhighlight a removed metadata field
-fix the collapse oddness

dstopp committed:
-update title
-minor spacing changes
-disable remove if there is only one <css part>

dstopp committed:
hide the dropdown on add

dstopp committed:
- add a loader on 'done'
- overlay the 'add' button on top of the data

dstopp committed:
- add a loader on 'done'
- overlay the 'add' button on top of the data

dstopp committed:
-overlay actually overlays on edit;
-fixed add

dstopp committed:
-modify metadata mini-menu to only show those not yet selected;
-add a flag to show currently editting so on update, the edit mode is not lost

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 417_reorder_table_column

# Conflicts:
#	opus/application/static_media/js/browse.js

dstopp committed:
-rename the dropdown to add_field <from minimenu>
-move the dropdown into the galleryView container so that its location can be offset relative to that container instead of body
-if the menu is already showing and the user clicks on a different 'Add field here', close current dropdown and reopen on new location

dstopp committed:
-moved the edit line out of the container so that it doesn't scroll
-update the percent of height used by the contents
-add no-select to the edit line and on edit of the list

dstopp committed:
-moved the dropdown a bit relative to target element;
-add rerender of select metadata modal; need to prevent screen selection until render complete <outstanding>

juzen2003 committed:
Fixed the op-add-metadata-fields dropdown menu jumping issue when clicking
categories.

dstopp committed:
-make the dd line a min height to account for empty data field
-fix for click on menu and have it close <should not close at that spot>

dstopp committed:
blasted typo

dstopp committed:
Merge pull request #7 from juzen2003/417_reorder_table_column

Fixed the op-add-metadata-fields dropdown menu jumping issue

juzen2003 committed:
Update PS metadata detailed modal, when any new detailed item is added.

dstopp committed:
Merge pull request #8 from juzen2003/417_reorder_table_column

Update PS in metadata detailed modal, when any new detailed item is added.

dstopp committed:
-fix right margin
-remove underline on hover of Edit/Done

dstopp committed:
fix a race condition caused when the selectMetadata html comes in before the #search page is done rendering

dstopp committed:
-make the dropdown menu a constant height;
-push menu up if on lower half of the screen, down if upper half
-update both select metadata and menu on change of select metatdata
-move the disable of user interaction to after the select metadata has been updated

dstopp committed:
tie the select metadata modal collapse/expand to the add field on the slide view so the both collapse/expand

dstopp committed:
minor responsiveness update

dstopp committed:
-remove the 'flash'  in the data when the metadata fields are updated
-prevent mousedrag on the add field menu from closing and dragging the modal
-change the cursor for the reorder to grabbing so it's different

dstopp committed:
-added responsiveness to the mutation observer

dstopp committed:
-add small padding above the tools
-jsHint updates
-close menu and change from edit to done when previous/next clicked.

dstopp committed:
-fix the pointer on the field hand
-moving the keyboard left/right arrows cancel edit mode
-if the last item added at the bottom causes the scrollbar to appear, make the entire newly added element visible
-fix a race condition w/the user interaction disabled

dstopp committed:
replaced the dl/dt/dd w/divs to see if that might fix the sortable issue <no luck yet>

dstopp committed:
-fix issue w/thumb shrinking when edit button is clicked on narrow screen

dstopp committed:
more updates to resolve the scrolling issue <still under construction>

dstopp committed:
-fix for drag and sort issue

juzen2003 commented on opus/application/apps/ui/templates/ui/add_field.html:
```diff
@@ -0,0 +1,100 @@
+                            <span class="title">{{ div.label }}&nbsp;<span class="op-menu-arrow"></span></span>
+                        </a>
+                <ul class="list-group list-group-flush submenu collapse {{ div.show }}" id="{{ which }}-mini-{{ div.table_name }}">
+
```
The indentation of ```<a>``` here should be aligned with ```<ul >```.

dstopp replied:
done

juzen2003 commented on opus/application/apps/ui/templates/ui/add_field.html:
```diff
@@ -0,0 +1,100 @@
+                                                        {% else %}
+                                                            {{ p.body_qualified_label }}
+                                                        {% endif %}
+                                                      " data-slug="{{ p.slug }}" class="op-search-param" href="#">
```
Need 2 more spaces here to aligned with other codes, same at line 72 and 78.

dstopp replied:
done

dstopp committed:
move code related to column sorting for the time being

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -301,6 +304,91 @@ var o_browse = {
+        });
+
+        $("#galleryView").on("click", "a.op-metadata-detail-remove", function(e) {
+            let slug = $(this).parent().parent().data('slug');
```
Just a note here, we can also use .closest("ul") as the replacement for .parent().parent()

dstopp replied:
thanks, good suggestion.  done

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -2069,23 +2179,137 @@ var o_browse = {
+        let selectMetadataTitle = "Add metadata field after the current field";
+        let removeTool = `<li class="op-metadata-details-tools list-inline-item">` +
+                         `<a href="#" class="op-metadata-detail-remove" title="Remove selected metadata field"><i class="far fa-trash-alt"></i></a></li>`;
+        //let addTool = `<a href="#" class="op-metadata-details-tools op-metadata-detail-add dropdown" title="${selectMetadataTitle}" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" "><i class="fas fa-plus pr-1"> Add field here</i></a>`;
```
Please remove the commented out variable (not used).

dstopp replied:
done

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -301,6 +304,91 @@ var o_browse = {
+                $("#op-add-metadata-fields").data("last", slug);
+                $(`.op-metadata-details > .loader`).show();
+                $((`#op-add-metadata-fields .op-select-list a[data-slug="${slug}"]`)).hide();
+                //o_menu.markMenuItem(`#op-add-metadata-fields .op-select-list a[data-slug="${slug}"]`);
```
Please remove unused function call.

dstopp replied:
done

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -301,6 +304,91 @@ var o_browse = {
+
+        $(".op-select-list").on("click", ".dropdown-item", function(e) {
+            o_browse.hideMenus();
+            //o_sortMetadata.onClickSortOrder($(this).data("slug"));
```
Please remove unused function call.

dstopp replied:
done

juzen2003 reviewed (changes requested):
- Database used: opus3_test_200517
- All Django tests pass: Y
- JSHINT run on all affected files: Y
- Tested on Chrome, Firefox, Safari, and iOS: Y
  (Remember to test all browser sizes)
- Tested for race conditions (with delayed API calls if applicable): Y

- Issues:
    - In metadata detail edit menu, when dragging around an item, release and quickly drag any item again, there will be some console errors. Here are steps to reproduce:
        1. In edit detail menu, drag an item, and move around.
        2. Release the item at any position (by releasing the mouse), and quickly hit the mouse again to drag any item.
        3. Move around for some time and the console will have the following errors from jquery ui:
            ```
            Uncaught TypeError: Cannot read property 'insertBefore' of null
            Uncaught TypeError: Cannot read property 'removeChild' of null
            ```
        (Issue can be reproduced by repeating the above steps)
    - In metadata detail slide show, when screen is small, metadata detail info will overlap a little bit with the bottom row (the row of cart, prev/next, and hamburger icons).

dstopp committed:
-address formatting issues noted in code review;
-fix a minor race condition in the sorting, where sorting can start on a new object before the previous sort is complete

juzen2003 committed:
Dynamically update the height of "#galleryViewContents .op-metadata-details"
to make sure items in the right columns of metadata detail modal will not
overlap with the bottom row.

juzen2003 committed:
Suppress x-scrolling in metadata detail modal.

juzen2003 committed:
Add comment at line 1112-1113 in opus.css.

juzen2003 committed:
Add styling to make sure info with long string like "Multiple Target List"
is wrapped in metadata detail modal.

dstopp committed:
Merge pull request #9 from juzen2003/417_reorder_table_column

417 reorder table column

dstopp committed:
-fix for offset over the tools when horizonally shrunk
-fix for console messages when the sort happens fast

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 417_reorder_table_column

rfrenchseti commented on opus/application/static_media/js/utils.js:
```diff
@@ -10,6 +10,8 @@
 var o_utils = {
 /* jshint varstmt: true */
 
+    numberOfDisableUserInteractionRequests: 0,
```
This variable is never used anywhere.

dstopp replied:
done

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1235,6 +1326,10 @@ var o_browse = {
+        let hashArray = o_hash.getHashArray();
+        let slugs = hashArray.cols.split(",");
+
+
```
Do you want two blank lines here?

dstopp replied:
done

rfrenchseti reviewed (changes requested):
- If you resize the browser while the dropdown is open, you can easily make it run off the top or bottom of the browser. I think it would make more sense if on resize we just closed the dropdown.
- As soon as you start dragging a field, a "hole" opens up below the field and all the lower fields move down one slot. This is not how dragging anywhere else works.
- Because of the above, you can end up with a weird visual like shown below (note duplicated field name).

![Screenshot from 2020-07-21 17-33-16](https://user-images.githubusercontent.com/6488118/88120690-62a95500-cb78-11ea-8ad7-540a45d90dde.png)

- Once all of the fields in a category are selected, the category name is still present. Wouldn't it be better to remove the category name from the list entirely? The way it is right now you just keep clicking on the category name expecting something to happen, and it never does. Don't forget to handle the case where there are no fields left to add.
- I think there needs to be a little more vertical space between "Edit" and the first metadata line. Perhaps you could just move "Edit" up slightly and leave everything else where it is.
- If you have a very wide field contents, the default browser X scrollbar will show up. I thought Dave had fixed this - he and I talked about it yesterday.

![Screenshot from 2020-07-21 17-39-04](https://user-images.githubusercontent.com/6488118/88120939-17437680-cb79-11ea-8729-3dc7fbc00c6c.png)

- I thought you were going to disable interaction while the new data was loading after adding/dragging/removing a field? Right now there's a race condition where if you add a field, and then add a new field while the previous adding is processing in the backend, only the final field will get added and the earlier ones will be thrown away. It's easy to reproduce with backend API delays.

dstopp committed:
-fix indenting in add_field
-close the add menu on resize of the browser
-when sorting the metadata order, the 'hole' is gone and the behaviour looks like the other sort things now
-if all the fields of a category are already selected, remove the empty category from the add list
-add a little space between "Edit" and the first metadata line
-removed unused variable
-remove extra blank line

dstopp committed:
-make the height of the placeholder in the detail sorted list 0 so that there is no 'hold'
-set no-select for the dropdown menus <tested cntl-A as well>
-fix the wrap on long field data

dstopp committed:
-fix debounce issue which caused crash on next/prev and other things
-rerender the menu after a change on selectMetadata modal

dstopp committed:
-fix empty data spacing
-add/remove pointer events on modals <didn't propagate from body>
-add spinners to the galleryView modal when the metadata fields change
-fix the bounce & rename function to be consistant
-typo in building of metadata fields render

juzen2003 committed:
Update the placeholder height in the start event of jquery ui sortable
on detail metadata edit menu.

dstopp committed:
Merge pull request #10 from juzen2003/417_reorder_table_column

Update the placeholder height in the start event of jquery ui sortable

dstopp committed:
rename variable

dstopp committed:
Merge branch '417_reorder_table_column' of https://github.com/dstopp/pds-opus into 417_reorder_table_column

dstopp committed:
-if all the metadata fields are used, disable the add field buttons
-indent the subheadings in the submenus
-if all the submenu fields are used, make sure the submenu header is gone
-disable the left/right arrow while the sort is active
-remove unnecessary update when edit is complete

dstopp committed:
-removed the debounce
-create a global variable for disabling the right/left arrow keys
-prevent pop up of the slide/gallery view until page is loaded

dstopp committed:
-minor jsHint
-added back some removed code for the close dialog
